src/GafferScene/Instancer.cpp

```diff
@@ -2650,39 +2650,116 @@ void Instancer::InstancerCapsule::render( IECoreScenePreview::Renderer *renderer
+		tbb::spin_rw_mutex prototypesMutex;
 
-			engines[0]->setPrototypeContextVariables( pointIndex, prototypeScope );
-			auto inserted = prototypeContextsMap.emplace( prototypeScope.context()->hash(), prototypeContextsMap.size() );
-
-			if( inserted.second )
+		task_group_context taskGroupContext( task_group_context::isolated );
+		tbb::parallel_for( tbb::blocked_range<size_t>( 0, pointIndicesForPrototype.size() ), [&]( const tbb::blocked_range<size_t> &r )
 			{
-				prototypes.push_back(
-					addPrototypeToVectors(
-						prototypesPlug, prototypeRoot, prototypeScope, sampleTimes, capsuleHash, *renderOptions
-					)
-				);
-			}
+				Context::EditableScope prototypeScope( scope.context() );
+				for( size_t i = r.begin(); i != r.end(); i++ )
+				{
+					int pointIndex = pointIndicesForPrototype[i];
+
+					// We find the capsules using the engine at shutter open, but the time used to construct the capsules
+					// must be the on-frame time, since the capsules will add their own shutter ( and we also handle
+					// the shutter ourselves for transform matrices )
+					//
+					// For most context variables, we are overwriting them for each prototype anyway, so
+					// we can reuse the context. But timeOffset is relative, so it's important that we reset the
+					// time before we do setPrototypeContextVariables for the next element. ( Should this be more
+					// general instead of assuming that frame is the only variable for which offsetMode may be set? )
+					prototypeScope.setFrame( onFrameTime );
+
+					engines[0]->setPrototypeContextVariables( pointIndex, prototypeScope );
+
+					MurmurHash contextHash = prototypeScope.context()->hash();
+
+					// We only need a read lock to check if there is already a prototype corresponding to this
+					// context.
+					tbb::spin_rw_mutex::scoped_lock prototypeContextsLock( prototypeContextsMapMutex, false );
+					size_t prototypeIndex;
+					auto existingPrototypeIndex = prototypeContextsMap.find( contextHash );
+					if( existingPrototypeIndex != prototypeContextsMap.end() )
+					{
+						// Found an existing prototype, we just use this index
+						prototypeIndex = existingPrototypeIndex->second;
+						prototypeContextsLock.release();
+					}
+					else
+					{
+						// Need to add a new prototype, so we need a write lock on the contexts map to
+						// add a new index
+						prototypeContextsLock.upgrade_to_writer();
+						auto inserted = prototypeContextsMap.emplace( contextHash, prototypeContextsMap.size() );
 
-			prototypeIndices.push_back( inserted.first->second );
+						prototypeIndex = inserted.first->second;
 
-		}
+						// We now have an index, and no longer need to access the contexts map.
+						prototypeContextsLock.release();
+
+						if( !inserted.second )
+						{
+							// Someone else allocated a prototype for this context while we were acquiring the
+							// write lock, we don't need to add anything after all
+						}
```

Comment:
I'm trying to reason a bit about the performance characteristics of this approach of computing prototypes upfront in a separate `parallel_for` at the start, vs doing them on the fly in the final loop where we actually output instances. One potential benefit seems to be here where we don't even need to wait for the prototype to be ready - we're happy to just get an index, and then we can carry on and process a new point. But there seem to be potential downsides as well, which I'll note in a separate comment below.